proposals/extensions_v2.md

```diff
@@ -0,0 +1,401 @@
+
+    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
+    // note: no `Assert val`, a static method is not passed an instance value.
+    public static bool Boo(string s1, string s2) for Assert { }
+
+    // Static extensions properties, indexers and events are all conceptually supportable.
+    // Though we can decide which are sensible to have or not.
+    // Static extensions can having backing static fields. Static extension properties can use `field`.
+    
+    // Nested types must be supported to maintain compatibility with existing static classes with extension members in them.
+}
+```
+
+Without specifying the full grammar changes, the intuition is that we are making the following changes:
+
+```g4
+for-clause
+    | 'for' parameter
+    ;
+
+parameter (no change)
+    | attributes? modifiers? type identifier ...
+    ;
+
+extension
+    | attributes? modifiers? 'extension' identifier { member_declaration* }
+    ;
+
+// For method/property/indexer/operator/constructor/event declarations
+// we are augmenting its syntax to allow type-parameters
+// (if not already allowed) and a for-clause. For example:
+property-declaration
+    | attributes? modifiers identifier type-parameters for-clause property-body
+    ;
+    
+compilation-unit-member
+    | ...
+    | extension
+    ;
+ 
+namespace-declaration-member
+    | ...
+    | extension
+    ;
+```
+
+The use of `parameter` means all of the following are legal, with the same semantics that that classic extension methods have today:
+
+```c#
+for ref Span<T> span
+for ref readonly Span<T> span
+for in Span<T> span
+for scoped ref Span<T> span
+for scoped Span<T> span
+```
+
+Modern extensions continue to not allow adding fields or destructors to a type.
+
+Not all of these *new* extension member forms need be supported.  For example, we may decide that a `static extension indexer` or `static extension event` is just too esoteric, and can be cut.  The proposal shows them all though to demonstrate completeness of the idea.  All *classic* forms must be supported of course.
+
+
+## Migration and compatibility
+
+Given an existing static class with extensions, a straightforward *semantically identical* (both at the source and binary level) translation to modern extensions is done in the following fashion.
+
+```c#
+// Existing style
+static class E
+{
+    static TField field;
+    static int Property => ...
+    static void NonExtensionHelperMethod() { }
+
+    static int ExtensionMethod(this string x, ...) { }
+    static T GenericExtensionMethod<T, U>(this U u, ...) { }
+}
+
+// New style
+extension E
+{
+    // Non extensions stay exactly the same.
+    static TField field;
+    static int Property => ...
+
+    // Note the lack of a 'for-clause'.  This is a normal static method.
+    // An *modern static extension method* will have a 'for-clause' on it
+    static void NonExtensionHelperMethod() { }
+
+    // Migrated *instance* extension members
```

Comment:
Classic extension methods could also be allowed inside `extension`. `this` parameter on a `static` method and `for` on an instance method would be strictly equivalent.

Reply:
They could be. But the point of this proposal is the *minimal* core. 

It would be fine to have an augmentation to say you could still write old -styles in the new extensions.